eval/narrow.elv

```diff
@@ -1,78 +1,124 @@
+	      ]
+  })]
 
-    edit:-narrow-read {
-        put $@candidates
-    } {
-        cd $0[content]
-    } &modeline="Location "
+  edit:-narrow-read {
+    put $@candidates
+  } [arg]{
+    cd $arg[content]
+    for hook $after-location { $hook }
+  } &modeline="[narrow] Location " &ignore-case=$true
 }
 
 fn history {
-    candidates = [(edit:command-history | each {
+  for hook $before-history { $hook }
+  candidates = [(edit:command-history | each [arg]{
         put [
-	    &content=$0[cmd]
-	    &display=$0[id]" "$0[cmd]
+	        &content=$arg[cmd]
+	        &display=$arg[id]" "(replaces "\t" " " (replaces "\n" " " $arg[cmd]))
         ]
-    })]
+  })]
 
-    edit:-narrow-read {
-        put $@candidates
-    } {
-        edit:replace-input $0[content]
-    } &modeline="History " &keep-bottom=$true
+  edit:-narrow-read {
+    put $@candidates
+  } [arg]{
+    edit:replace-input $arg[content]
+    for hook $after-history { $hook }
+  } &modeline="[narrow] History " &keep-bottom=$true &ignore-case=$true
 }
 
 fn lastcmd {
-    last = (edit:command-history -1)
-    cmd = [
-            &content=$last[cmd]
-            &display="-1 "$last[cmd]
-	    &filter-text=""
-        ]
-    index = 0
-    candidates = [$cmd ( edit:wordify $last[cmd] | each {
-	put [
-            &content=$0
-            &display=$index" "$0
-            &filter-text=$index
-	]
-	index = (+ $index 1)
-    })]
-    edit:-narrow-read {
-        put $@candidates
-    } {
-        edit:replace-input $0[content]
-    } &modeline="Lastcmd " &auto-commit=$true &bindings=[&M-1={ edit:narrow:accept-close }]
+  for hook $before-lastcmd { $hook }
+  last = (edit:command-history -1)
+  cmd = [
+    &content=$last[cmd]
+    &display="M-1 "$last[cmd]
+	  &filter-text=""
+  ]
+  index = 0
+  candidates = [$cmd ( edit:wordify $last[cmd] | each [arg]{
+	      put [
+          &content=$arg
+          &display=$index" "$arg
+          &filter-text=$index
+	      ]
+	      index = (+ $index 1)
+  })]
+  edit:-narrow-read {
+    put $@candidates
+  } [arg]{
+    edit:insert-at-dot $arg[content]
+    for hook $after-lastcmd { $hook }
+  } &modeline="[narrow] Lastcmd " &auto-commit=$true &bindings=[&M-1={ edit:narrow:accept-close }] &ignore-case=$true
 }
 
 
-# TODO: separate bindings from functions
+fn -bind_i [k f]{
+  edit:insert:binding[$k] = $f
+}
 
-fn bind [m k f]{
-    edit:binding[$m][$k] = $f
+fn -bind_n [k f]{
+  edit:narrow:binding[$k] = $f
 }
 
-bind insert C-l       narrow:location
-bind insert C-r       narrow:history
-bind insert M-1       narrow:lastcmd
+# Bind keys for location, history and lastcmd modes. Without
+# options, it uses the default bindings, but different keys
+# can be specified with the options. To disable a binding,
+# specify its key as "".
+# Example:
+#   narrow:bind_keys &location=Alt-l &lastcmd=""
+fn bind_keys [&location=C-l &history=C-r &lastcmd=M-1]{
+  if (> (count $location) 0) { -bind_i $location narrow:location }
```

Comment:
I was going to say `if not-eq $location ""` is clearer here, then I found out that we don't have that `not-eq` builtin. I just added it, so that I can now say `if not-eq $location ""` is clearer. :)